raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1555,28 +1550,65 @@ public void poll() throws IOException {
-                    unwrittenAppend.complete(offsetAndEpoch);
-                    uncommittedAppends.put(offsetAndEpoch, new AppendBatchAndTime(info.lastOffset - info.firstOffset + 1, currentTimeMs));
+            int count = 0;
+            while (!unwrittenAppends.isEmpty() && count < MAX_RECORDS_DRAINED_PER_ITERATION) {
+                final UnwrittenAppend unwrittenAppend = unwrittenAppends.poll();
+
+                if (unwrittenAppend.isCancelled())
+                    continue;
+
+                if (unwrittenAppend.isTimedOut(currentTimeMs)) {
+                    unwrittenAppend.fail(new TimeoutException());
                 } else {
-                    unwrittenAppend.fail(new InvalidRequestException("Leader refused the append"));
+                    LeaderState leaderState = quorum.leaderStateOrThrow();
+                    int epoch = quorum.epoch();
+                    LogAppendInfo info = maybeAppendAsLeader(leaderState, unwrittenAppend.records);
+                    if (info != null) {
+                        OffsetAndEpoch offsetAndEpoch = new OffsetAndEpoch(info.lastOffset, epoch);
+
+                        if (unwrittenAppend.ackMode == AckMode.LEADER) {
+                            unwrittenAppend.complete(offsetAndEpoch);
+                        } else if (unwrittenAppend.ackMode == AckMode.QUORUM) {
+                            CompletableFuture<Long> future = new CompletableFuture<>();
+                            appendPurgatory.await(future, offsetAndEpoch.offset, unwrittenAppend.requestTimeoutMs);
+
+                            future.whenComplete((completionTimeout, exception) -> {
+                                if (exception != null) {
+                                    logger.error("Failed to commit append at {} due to {}", offsetAndEpoch, exception);
+
+                                    unwrittenAppend.fail(exception);
+                                } else if (!quorum.isLeader()) {
+                                    // if we are no longer the leader, fail the future instead
+                                    unwrittenAppend.fail(new NotLeaderOrFollowerException("Failed to receive " +
+                                            "sufficient acknowledgments for this append before leader change."));
+                                } else {
+                                    long elapsedTime = Math.max(0, completionTimeout - currentTimeMs);
+                                    long records = info.lastOffset - info.firstOffset + 1;
+                                    double elapsedTimePerRecord = (double) elapsedTime / records;
+                                    kafkaRaftMetrics.updateCommitLatency(elapsedTimePerRecord, currentTimeMs);
+                                    unwrittenAppend.complete(offsetAndEpoch);
+
+                                    logger.debug("Completed committing append with {} records at {}", records, offsetAndEpoch);
```

Comment:
Is the intent to print all the records or a count?

Reply:
The intent is to print just the count.

Reply:
nit: maybe rename to `numRecords`